plugin/src/main/java/org/opensearch/ml/rest/RestStatsMLAction.java

```diff
@@ -58,60 +84,147 @@ public List<Route> routes() {
+            ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser);
+            mlStatsInput = MLStatsInput.parse(parser);
+        } else {
+            mlStatsInput = createMlStatsInputFromRequestParams(request);
+        }
 
-    /**
-     * Creates a MLStatsNodesRequest from a RestRequest
-     *
-     * @param request RestRequest
-     * @return MLStatsNodesRequest
-     */
-    @VisibleForTesting
-    MLStatsNodesRequest getRequest(RestRequest request) {
-        // todo: add logic to triage request based on node type(ML node or data node)
-        MLStatsNodesRequest mlStatsRequest = new MLStatsNodesRequest(splitCommaSeparatedParam(request, "nodeId").orElse(null));
-        mlStatsRequest.timeout(request.param("timeout"));
+        String[] nodeIds = mlStatsInput.retrieveStatsOnAllNodes() ? getAllNodes() : mlStatsInput.getNodeIds().toArray(new String[0]);
+        MLStatsNodesRequest mlStatsNodesRequest = new MLStatsNodesRequest(nodeIds, mlStatsInput);
 
-        List<String> requestedStats = splitCommaSeparatedParam(request, "stat").map(Arrays::asList).orElseGet(Collections::emptyList);
+        Map<MLClusterLevelStat, Object> clusterStatsMap = new HashMap<>();
+        if (mlStatsInput.getTargetStatLevels().contains(MLStatLevel.CLUSTER)) {
+            clusterStatsMap.putAll(getClusterStatsMap(mlStatsInput));
+        }
+
+        // copy mlStatsInput to make an effectively final temp variable finalMlStatsInput
+        MLStatsInput finalMlStatsInput = mlStatsInput;
+        return channel -> {
+            if (finalMlStatsInput.getTargetStatLevels().contains(MLStatLevel.CLUSTER)
+                && (finalMlStatsInput.retrieveAllClusterLevelStats()
+                    || finalMlStatsInput.getClusterLevelStats().contains(MLClusterLevelStat.ML_MODEL_COUNT))) {
+                indexUtils.getNumberOfDocumentsInIndex(ML_MODEL_INDEX, ActionListener.wrap(count -> {
+                    clusterStatsMap.put(MLClusterLevelStat.ML_MODEL_COUNT, count);
+                    getNodeStats(finalMlStatsInput, clusterStatsMap, client, mlStatsNodesRequest, channel);
+                }, e -> {
+                    String errorMessage = "Failed to get ML model count";
+                    log.error(errorMessage, e);
+                    onFailure(channel, RestStatus.INTERNAL_SERVER_ERROR, errorMessage, e);
+                }));
+            } else {
+                getNodeStats(finalMlStatsInput, clusterStatsMap, client, mlStatsNodesRequest, channel);
+            }
+        };
+    }
 
-        Set<String> validStats = mlStats.getStats().keySet();
-        if (isAllStatsRequested(requestedStats)) {
-            mlStatsRequest.setRetrieveAllStats(true);
+    MLStatsInput createMlStatsInputFromRequestParams(RestRequest request) {
+        MLStatsInput mlStatsInput = new MLStatsInput();
+        Optional<String[]> nodeIds = splitCommaSeparatedParam(request, "nodeId");
```

Comment:
This is to parse all the nodeIds in the url in "GET /_plugins/_ml/{nodeId}/stats"? For example, GET /_plugins/_ml/nodeId1,nodeId2/stats would return stats for both id1 and id2?

Reply:
Correct, user can get multiple node's stats by separating node ids with comma.

Reply:
Is this going to raise any security concern? Like can a person call this on a node that they doesn't own?